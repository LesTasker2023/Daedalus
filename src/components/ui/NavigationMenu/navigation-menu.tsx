import { Menu, X } from 'lucide-react'
import React from 'react'
import {
    NavigationMenu as NavigationMenuBase,
    NavigationMenuContent,
    NavigationMenuIndicator,
    NavigationMenuItem,
    NavigationMenuLink,
    NavigationMenuList,
    NavigationMenuTrigger,
    NavigationMenuViewport,
} from '../navigation-menu'

interface NavigationMenuProps {
  mobile?: boolean
  loading?: boolean
  children: React.ReactNode
  className?: string
}

const NavigationMenu = ({ mobile, loading, children, className, ...props }: NavigationMenuProps) => {
  const [isMobileMenuOpen, setIsMobileMenuOpen] = React.useState(false)

  if (loading) {
    return (
      <div className="flex items-center justify-center p-4">
        <div className="animate-spin w-6 h-6 border-2 border-gray-300 border-t-gray-600 rounded-full" />
        <span className="ml-2 text-sm text-muted-foreground">Loading menu...</span>
      </div>
    )
  }

  if (mobile) {
    return (
      <div className="relative">
        {/* Mobile Toggle Button */}
        <button
          onClick={() => setIsMobileMenuOpen(!isMobileMenuOpen)}
          className="md:hidden p-2 rounded-md hover:bg-gray-100"
          aria-label="Toggle navigation menu"
        >
          {isMobileMenuOpen ? <X className="w-6 h-6" /> : <Menu className="w-6 h-6" />}
        </button>

        {/* Desktop Navigation (hidden on mobile) */}
        <div className="hidden md:block">
          <NavigationMenuBase className={className} {...props}>
            {children}
          </NavigationMenuBase>
        </div>

        {/* Mobile Navigation Menu */}
        {isMobileMenuOpen && (
          <>
            {/* Overlay */}
            <div 
              className="navigation-menu__mobile-overlay fixed inset-0 bg-black/50 z-40 md:hidden"
              onClick={() => setIsMobileMenuOpen(false)}
            />
            
            {/* Mobile Menu */}
            <div className="fixed top-0 left-0 right-0 bg-white border-b shadow-lg z-50 md:hidden">
              <div className="flex items-center justify-between p-4 border-b">
                <h2 className="text-lg font-semibold">Navigation</h2>
                <button
                  onClick={() => setIsMobileMenuOpen(false)}
                  className="p-1 rounded-md hover:bg-gray-100"
                >
                  <X className="w-5 h-5" />
                </button>
              </div>
              <div className="p-4">
                <NavigationMenuBase className="w-full">
                  {children}
                </NavigationMenuBase>
              </div>
            </div>
          </>
        )}
      </div>
    )
  }

  return (
    <NavigationMenuBase className={className} {...props}>
      {children}
    </NavigationMenuBase>
  )
}

// Add compound component pattern
NavigationMenu.List = NavigationMenuList
NavigationMenu.Item = NavigationMenuItem
NavigationMenu.Trigger = NavigationMenuTrigger
NavigationMenu.Content = NavigationMenuContent
NavigationMenu.Link = NavigationMenuLink
NavigationMenu.Indicator = NavigationMenuIndicator
NavigationMenu.Viewport = NavigationMenuViewport

export default NavigationMenu
export {
    NavigationMenu, NavigationMenuContent, NavigationMenuIndicator, NavigationMenuItem, NavigationMenuLink, NavigationMenuList, NavigationMenuTrigger, NavigationMenuViewport
}

