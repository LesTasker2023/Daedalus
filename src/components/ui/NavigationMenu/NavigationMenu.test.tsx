import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { NavigationMenu } from './navigation-menu';

describe('NavigationMenu', () => {
  const renderBasicNavigationMenu = (props = {}) => {
    return render(
      <NavigationMenu data-testid="navigationmenu" {...props}>
        Test content
      </NavigationMenu>
    );
  };

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = renderBasicNavigationMenu();
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Basic Functionality', () => {
    it('renders correctly', () => {
      renderBasicNavigationMenu();
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it.skip('can be focused - SKIPPED: Non-focusable element', () => {
      // This element cannot receive focus (div/span/table)
      // Focus tests disabled for accessibility accuracy
      expect(true).toBe(true);
    });

    it('has proper ARIA attributes', () => {
      renderBasicNavigationMenu();
      const element = screen.getByTestId('navigationmenu');
      expect(element).toBeInTheDocument();
      // TODO: Add specific ARIA attribute tests based on component type
    });

    it.skip('supports keyboard navigation - SKIPPED: Non-focusable element', () => {
      // This element cannot receive focus (div/span/table)
      // Focus tests disabled for accessibility accuracy
      expect(true).toBe(true);
    });

    it('announces changes to screen readers', () => {
      renderBasicNavigationMenu();
      // TODO: Add screen reader announcement tests
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });

    it('respects reduced motion preferences', () => {
      renderBasicNavigationMenu();
      // TODO: Add reduced motion tests
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });
  });

  describe('Custom Styling and Props', () => {
    it('accepts custom className', () => {
      renderBasicNavigationMenu({ className: 'custom-class' });
      const element = screen.getByTestId('navigationmenu');
      expect(element).toHaveClass('custom-class');
    });

    it('forwards refs correctly', () => {
      const ref = vi.fn();
      renderBasicNavigationMenu({ ref });
      // Ref forwarding test - environment dependent
      // expect(ref).toHaveBeenCalledWith(expect.any(HTMLElement));
    });

    it('spreads additional props', () => {
      renderBasicNavigationMenu({ 'data-custom': 'test-value' });
      const element = screen.getByTestId('navigationmenu');
      expect(element).toHaveAttribute('data-custom', 'test-value');
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined props gracefully', () => {
      renderBasicNavigationMenu({ children: undefined });
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });

    it('handles null props gracefully', () => {
      renderBasicNavigationMenu({ children: null });
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });

    it('handles empty string props', () => {
      renderBasicNavigationMenu({ className: '' });
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });

    it('handles rapid prop changes', () => {
      const { rerender } = renderBasicNavigationMenu({ className: 'class1' });
      rerender(<NavigationMenu data-testid="navigationmenu" className="class2" />);
      const element = screen.getByTestId('navigationmenu');
      expect(element).toHaveClass('class2');
    });

    it('handles complex nested content', () => {
      render(
        <NavigationMenu data-testid="navigationmenu">
          <div>
            <span>Nested content</span>
            <p>More content</p>
          </div>
        </NavigationMenu>
      );
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });

    it('maintains functionality with many children', () => {
      render(
        <NavigationMenu data-testid="navigationmenu">
          {Array.from({ length: 100 }, (_, i) => (
            <div key={i}>Item {i}</div>
          ))}
        </NavigationMenu>
      );
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });

    it('handles component unmounting cleanly', () => {
      const { unmount } = renderBasicNavigationMenu();
      expect(() => unmount()).not.toThrow();
    });

    it('preserves functionality after remounting', () => {
      const { unmount } = renderBasicNavigationMenu();
      unmount();
      renderBasicNavigationMenu();
      expect(screen.getByTestId('navigationmenu')).toBeInTheDocument();
    });
  });
});

// TODO: Review and customize generated tests based on component-specific requirements
// TODO: Add component-specific interaction tests
// TODO: Verify all variant combinations work correctly
// TODO: Test integration with form libraries if applicable
// TODO: Add performance tests for complex components
