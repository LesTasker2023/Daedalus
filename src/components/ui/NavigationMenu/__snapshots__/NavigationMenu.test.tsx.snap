// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationMenu > Snapshots > matches default snapshot 1`] = `
<nav
  aria-label="Main"
  class="relative z-10 flex max-w-max flex-1 items-center justify-center"
  data-orientation="horizontal"
  data-testid="navigationmenu"
  dir="ltr"
>
  Test content
  <div
    class="absolute left-0 top-full flex justify-center"
  />
</nav>
`;
